compiler-core/src/build/package_loader.rs

```diff
@@ -1715,3 +1682,143 @@ impl<'a> Inputs<'a> {
+
+        // my/module
+        let _ = module_path.set_extension("");
+
+        // Stringify
+        let name = module_path.to_string();
+
+        // normalise windows paths
+        name.replace("\\", "/").into()
+    }
+
+    fn is_gleam_path(path: &Utf8Path, dir: &Utf8Path) -> bool {
+        use regex::Regex;
+        use std::cell::OnceCell;
+        const RE: OnceCell<Regex> = OnceCell::new();
+
+        RE.get_or_init(|| {
+            Regex::new(&format!(
+                "^({module}{slash})*{module}\\.gleam$",
+                module = "[a-z][_a-z0-9]*",
+                slash = "(/|\\\\)",
+            ))
+            .expect("is_gleam_path() RE regex")
+        })
+        .is_match(
+            path.strip_prefix(dir)
+                .expect("is_gleam_path(): strip_prefix")
+                .as_str(),
+        )
+    }
+}
+
+/// The collection of cache files paths related to a module.
+/// These files are not guaranteed to exist.
+pub struct CacheFiles {
+    pub cache_path: Utf8PathBuf,
+    pub meta_path: Utf8PathBuf,
+    pub warnings_path: Utf8PathBuf,
+}
+
+impl CacheFiles {
+    pub fn new(artefact_directory: &Utf8Path, module_name: &EcoString) -> Self {
+        let file_name = module_name.replace("/", "@");
+        let cache_path = artefact_directory
+            .join(file_name.as_str())
+            .with_extension("cache");
+        let meta_path = artefact_directory
+            .join(file_name.as_str())
+            .with_extension("cache_meta");
+        let warnings_path = artefact_directory
+            .join(file_name.as_str())
+            .with_extension("cache_warnings");
+
+        Self {
+            cache_path,
+            meta_path,
+            warnings_path,
+        }
+    }
+
+    pub fn delete(&self, io: &dyn io::FileSystemWriter) {
+        // TODO: Should we check for errors here?
+        let _ = io.delete_file(&self.cache_path);
+        let _ = io.delete_file(&self.meta_path);
+        let _ = io.delete_file(&self.warnings_path);
+    }
+
+    /// Iterates over `.cache_meta` files in the given directory,
+    /// and returns the respective module names.
+    /// Symlinks are followed.
+    pub fn modules_with_meta_files<'a>(
+        io: &'a impl FileSystemReader,
+        dir: &'a Utf8Path,
+    ) -> impl Iterator<Item = EcoString> + 'a {
+        tracing::trace!("CacheFiles::modules_with_meta_files {:?}", dir);
+        files_with_extension(io, dir, "cache_meta").map(move |path| Self::module_name(&dir, &path))
+    }
+
+    fn module_name(dir: &Utf8Path, path: &Utf8Path) -> EcoString {
+        // /path/to/artefact/dir/my@module.cache_meta
+
+        // my@module.cache_meta
+        let mut module_path = path
+            .strip_prefix(dir)
+            .expect("Stripping package prefix from module path")
+            .to_path_buf();
+
+        // my@module
+        let _ = module_path.set_extension("");
+
+        // my/module
+        module_path.to_string().replace("@", "/").into()
```

Comment:
This is the corrected and modified version of getting the module name from any cache file.